packages/app/src/systems/Core/components/CoinInput.tsx

```diff
@@ -77,3 +63,43 @@ export const CoinInput = forwardRef<HTMLInputElement, CoinInputProps>(
     );
   }
 );
+
+function useDisplayedCoins(
+  initialValue: string,
+  onChange: CoinInputProps["onChange"]
+): [string, (e: React.ChangeEvent<HTMLInputElement>) => void] {
+  const [value, setValue] = useState(initialValue);
+
+  useEffect(() => {
+    if (value !== initialValue) onChange?.(value);
+  }, [value]);
+
+  useEffect(() => {
+    if (parseFloat(value) === 0 && initialValue === ZERO_AMOUNT.value) return;
+    setValue(initialValue);
+  }, [initialValue]);
+
+  const valueSetter = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      if (e.currentTarget.value.startsWith(".")) {
+        setValue("0.");
+        return;
+      }
+
+      const hasLeadingZeros = e.currentTarget.value.match(/^0+\d/) != null;
+
+      if (hasLeadingZeros) {
+        const valueWithoutLeadingZeros = parseFloat(
```

Comment:
I think you can clean up the leading zeros direct by using the replace string. No need to parse to float as you already run a regex to check.